### PR TITLE
Implement computeResidualAndJacobian for a few non-AD objects

### PR DESCRIPTION
**Pull request by hugary1995**
Kernel, InterfaceKernel, NodalBC and IntegratedBC

ref #23487

### CODE REVIEW

**Comment by dschwen** [comment]
Ok, I suppose the benefit here is that we evaluate material properties once for both the residual and the Jacobian?

**Comment by hugary1995** [comment]
Yes. I briefly discussed that in the issue description (I should probably do that in the PR description as well). 

This strategy is attractive only when
1. the residual evaluation is approximately as expensive as the Jacobian evaluation;
2. we can avoid a lot of computation by evaluating residual and Jacobian together.

The "expensive material evaluation" is one scenario that satisfies the above two conditions. AD is another scenario apparently.

**Comment by hugary1995** [comment]
The caveat is that we need an additional Jacobian evaluation for the final convergence check if we compute residual and Jacobian together. So for example, suppose at a given time step there are N nonlinear iterations. If we compute residual and Jacobian separately, we only need N residual evaluations + (N-1) Jacobian evaluations. However, if we compute the residual and Jacobian together, we will need N "residual+Jacobian" evaluations. Therefore, this "residual_and_jacobian_together" option is more advantageous when N gets large. 

But how large does N has to be for this option to be preferrable? That depends on how much computational cost we can avoid by evaluation residual and Jacobian together. If the cost of one "residual+Jacobian" evaluation is as expensive as one Jacobian evaluation (for example, when the cost of material evaluation dominates), then this option will always be faster.

I'll add this discussion in the docs if you guys encourage this PR.

File: framework/src/bcs/NodalBC.C [hunk 1]
```diff
@@ -163,3 +163,27 @@ NodalBC::computeQpOffDiagJacobian(unsigned int /*jvar*/)
...
+  computeResidual();
+
+  auto & ce = _fe_problem.couplingEntries(_tid);
+  for (const auto & it : ce)
+  {
+    MooseVariableFieldBase & ivariable = *(it.first);
+    MooseVariableFieldBase & jvariable = *(it.second);
+
+    unsigned int ivar = ivariable.number();
+    unsigned int jvar = jvariable.number();
```
Comment (lindsayad) [inline framework/src/bcs/NodalBC.C:179]:
```suggestion
  for (const auto & [ivariable, jvariable] : ce)
  {
    const unsigned int ivar = ivariable->number();
    const unsigned int jvar = jvariable->number();
```

File: framework/src/constraints/MortarConstraintBase.C [hunk 2]
```diff
@@ -203,10 +203,9 @@ MortarConstraintBase::zeroInactiveLMDofs(const std::unordered_set<const Node *>
         continue;
 
       const auto dof_index = node->dof_number(sn, vn, 0);
-      if (_subproblem.currentlyComputingJacobian() ||
-          _subproblem.currentlyComputingResidualAndJacobian())
+      if (_assembly.computingJacobian())
```
Comment (lindsayad) [inline framework/src/constraints/MortarConstraintBase.C:206]:
```suggestion
      if (_assembly.computingJacobian() || _assembly.computingResidualAndJacobian())
```

File: framework/src/constraints/MortarConstraintBase.C [hunk 3]
```diff
@@ -203,10 +203,9 @@ MortarConstraintBase::zeroInactiveLMDofs(const std::unordered_set<const Node *>
...
         _assembly.cacheJacobian(dof_index, dof_index, 1., _matrix_tags);
-      if (!_subproblem.currentlyComputingJacobian())
+      if (_assembly.computingResidual())
```
Comment (lindsayad) [inline framework/src/constraints/MortarConstraintBase.C:208]:
```suggestion
      if (_assembly.computingResidual() || _assembly.computingResidualAndJacobian())
```

File: framework/src/bcs/IntegratedBC.C [hunk 4]
```diff
@@ -179,3 +179,30 @@ IntegratedBC::computeOffDiagJacobianScalar(const unsigned int jvar)
...
+  computeResidual();
+
+  auto & ce = _fe_problem.couplingEntries(_tid);
+  for (const auto & it : ce)
+  {
+    MooseVariableFieldBase & ivariable = *(it.first);
+    MooseVariableFieldBase & jvariable = *(it.second);
+
+    unsigned int ivar = ivariable.number();
+    unsigned int jvar = jvariable.number();
```
Comment (lindsayad) [inline framework/src/bcs/IntegratedBC.C:195]:
```suggestion
  for (const auto & [ivariable, jvariable] : ce)
  {
    const unsigned int ivar = ivariable->number();
    const unsigned int jvar = jvariable->number();
```

File: framework/src/interfacekernels/InterfaceKernel.C [hunk 5]
```diff
@@ -404,6 +404,34 @@ InterfaceKernelTempl<T>::computeNeighborOffDiagJacobian(unsigned int jvar)
...
+  if (!isImplicit())
+    return;
+
+  for (const auto & [ivariable, jvariable] : _fe_problem.couplingEntries(_tid))
+  {
+    if (ivariable->isFV())
+      continue;
+
+    unsigned int ivar = ivariable->number();
+    unsigned int jvar = jvariable->number();
```
Comment (lindsayad) [inline framework/src/interfacekernels/InterfaceKernel.C:422]:
```suggestion
    const unsigned int ivar = ivariable->number();
    const unsigned int jvar = jvariable->number();
```

File: framework/src/kernels/Kernel.C [hunk 6]
```diff
@@ -190,3 +190,30 @@ Kernel::computeOffDiagJacobianScalar(const unsigned int jvar)
...
   accumulateTaggedLocalMatrix();
 }
+
+void
+Kernel::computeResidualAndJacobian()
+{
+  computeResidual();
+
+  auto & ce = _fe_problem.couplingEntries(_tid);
+  for (const auto & it : ce)
```
Comment (lindsayad) [inline framework/src/kernels/Kernel.C:200]:
```suggestion
  for (const auto & [ivariable, jvariable] : ce)
```

File: framework/src/kernels/Kernel.C [hunk 7]
```diff
@@ -190,3 +190,30 @@ Kernel::computeOffDiagJacobianScalar(const unsigned int jvar)
...
+  {
+    MooseVariableFieldBase & ivariable = *(it.first);
+    MooseVariableFieldBase & jvariable = *(it.second);
+
+    unsigned int ivar = ivariable.number();
+    unsigned int jvar = jvariable.number();
```
Comment (lindsayad) [inline framework/src/kernels/Kernel.C:206]:
```suggestion
    const unsigned int ivar = ivariable.number();
    const unsigned int jvar = jvariable.number();
```

File: framework/src/constraints/MortarConstraintBase.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/constraints/MortarConstraintBase.C:206]: Ah I see what you did in `FEProblemBase.C`. Ignore

File: framework/src/constraints/MortarConstraintBase.C (hunk 3 above)
Comment (lindsayad) [reply framework/src/constraints/MortarConstraintBase.C:208]: Ah I see what you did in `FEProblemBase.C`. Ignore

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
Definitely will want more testing

**Comment by hugary1995** [comment]
The documentation in NonlinearSystem.md already does a good job at covering the case I discussed above. So I am not adding any. Let's see what the coverage report says now...

**Comment by hugary1995** [comment]
I suppose I should also clarify that I am changing the interpretation of "computing residual", "computing Jacobian", and "computing residual and Jacobian".

Prior to this PR, the three states are mutually exclusive. There was even an assertion that "computing Jacobian" and "computing residual and Jacobian" cannot both be true.

With this PR, I am redefining "computing residual and Jacobian" in the sense that when "computing residual and Jacobian" is true, both "computing residual" and "computing Jacobian" are true. However, "computing residual" == true or "computing Jacobian" == true does not imply that "computing residual and Jacobian" is true.

I find the new definition more intuitive and easier to handle programmatically. I am open to suggestions.

**Comment by lindsayad** [comment]
I am fine with the semantic changes, and I really appreciate that you've made the definitions of computing-residual, computing-jacobian, etc. consistent across `SubProblem` and `Assembly`. That is one I felt guilty about 😄

**Comment by hugary1995** [comment]
This is ready for another review. The rest of the civet failures are out of my control.

**Review by lindsayad — APPROVED** [review]
If I don't get any other reviews by the end of the day, then I will merge

**Pull request merged**